pkg/trace/writer/trace.go

```diff
@@ -127,183 +120,202 @@ func NewTraceWriter(
 	if s := cfg.TraceWriter.FlushPeriodSeconds; s != 0 {
 		tw.tick = time.Duration(s*1000) * time.Millisecond
 	}
+	tw.flushTicker = time.NewTicker(tw.tick)
+
 	qsize := 1
 	log.Infof("Trace writer initialized (climit=%d qsize=%d)", climit, qsize)
 	tw.senders = newSenders(cfg, tw, pathTraces, climit, qsize, telemetryCollector, statsd)
-	for i := 0; i < runtime.GOMAXPROCS(0); i++ {
-		tw.wg.Add(1)
-		go tw.serializer()
-	}
+	tw.wg.Add(1)
+	go tw.timeFlush()
+	tw.wg.Add(1)
+	go tw.reporter()
 	return tw
 }
 
-// Stop stops the TraceWriter and attempts to flush whatever is left in the senders buffers.
-func (w *TraceWriter) Stop() {
-	log.Debug("Exiting trace writer. Trying to flush whatever is left...")
-	w.stop <- struct{}{}
-	<-w.stop
-	// Wait for encoding/compression to complete on each payload,
-	// and submission to senders
-	w.wg.Wait()
-	stopSenders(w.senders)
-}
-
-// Run starts the TraceWriter.
-func (w *TraceWriter) Run() {
-	if w.syncMode {
-		w.runSync()
-	} else {
-		w.runAsync()
-	}
-}
-
-func (w *TraceWriter) runAsync() {
-	t := time.NewTicker(w.tick)
-	defer t.Stop()
-	defer close(w.Serialize)
-	defer close(w.stop)
+func (w *TraceWriter) reporter() {
+	tck := time.NewTicker(w.tick)
+	defer w.wg.Done()
 	for {
 		select {
-		case pkg := <-w.In:
-			w.addSpans(pkg)
+		case <-tck.C:
+			w.report()
 		case <-w.stop:
-			w.drainAndFlush()
 			return
-		case <-t.C:
-			w.report()
-			w.flush()
 		}
 	}
 }
 
-func (w *TraceWriter) runSync() {
-	defer close(w.Serialize)
-	defer close(w.stop)
-	defer close(w.flushChan)
+func (w *TraceWriter) timeFlush() {
+	defer w.wg.Done()
 	for {
 		select {
-		case pkg := <-w.In:
-			w.addSpans(pkg)
-		case notify := <-w.flushChan:
-			w.drainAndFlush()
-			notify <- struct{}{}
+		case <-w.flushTicker.C:
+			func() {
+				w.mu.Lock()
```

Comment:
nit: `FlushSync` function owns the mutex when flushing the buffer. Should we delegate the mutex to `flush` as well? It would reduce the logic around the mutex when calling `flush`